src/Microsoft.AspNetCore.Authorization.Policy/AuthorizationMiddleware.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authorization.Policy;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Authorization
+{
+    public class AuthorizationMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly IAuthorizationPolicyProvider _policyProvider;
+
+        public AuthorizationMiddleware(RequestDelegate next, IAuthorizationPolicyProvider policyProvider)
+        {
+            if (next == null)
+            {
+                throw new ArgumentNullException(nameof(next));
+            }
+
+            if (policyProvider == null)
+            {
+                throw new ArgumentNullException(nameof(policyProvider));
+            }
+
+            _next = next;
+            _policyProvider = policyProvider;
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var policyEvaluator = context.RequestServices.GetRequiredService<IPolicyEvaluator>();
+
+            var endpoint = context.Features.Get<IEndpointFeature>()?.Endpoint;
+            var authorizeData = endpoint?.Metadata.GetOrderedMetadata<IAuthorizeData>() ?? Array.Empty<IAuthorizeData>();
+            var policy = await AuthorizationPolicy.CombineAsync(_policyProvider, authorizeData);
+            if (policy == null)
+            {
+                await _next(context);
+                return;
+            }
+
+            var authenticateResult = await policyEvaluator.AuthenticateAsync(policy, context);
+
+            // Allow Anonymous skips all authorization
+            if (endpoint?.Metadata.GetMetadata<IAllowAnonymous>() != null)
+            {
+                await _next(context);
+                return;
+            }
+
+            // Note that the resource will be null if there is no matched endpoint
+            var authorizeResult = await policyEvaluator.AuthorizeAsync(policy, authenticateResult, context, resource: endpoint);
+
+            if (authorizeResult.Challenged)
+            {
+                if (policy.AuthenticationSchemes.Any())
+                {
+                    foreach (var scheme in policy.AuthenticationSchemes)
+                    {
+                        await context.ChallengeAsync(scheme);
```

Comment:
Should `ChallengeAsync` and `ForbidAsync` be passed the authentication properties from `authenticateResult`?

Reply:
I am not sure that makes sense, you likely got nothing back in the challenge case, and forbid means that was probably not the right identity.

The properties are more metadata associated with that particular result/request.  Things typically set in the properties for forbid/challenge are usually more like parameters to control behavior like urls

Reply:
The reason I brought it up is there is a bug in cookies auth around the redirect URL + app.Map - https://github.com/aspnet/AuthSamples/pull/65#issuecomment-435260693

I noticed that `ChallengeAsync` for cookies auth can read the `RedirectUri` from the authentication properties, so a fix for the bad redirect URL could be to hack in a new implementation of `IPolicyEvaluator.AuthenticateAsync` that returns a fixed `RedirectUri` in the result properties, and `ChallengeAsync` would then pick it up.

The bug is being fixed anyway, and this is a pretty esoteric scenario, just wondering if it should be done. My knowledge of ASP.NET Core security is high-level.

Reply:
No need, this challenge is instigating a new auth flow.

Reply:
ok!